### PR TITLE
Adding Detector Endpoints for Web Apps and App Service Environments

### PR DESCRIPTION
**Pull request by stevenernst131**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

There were errors and warnings that already exist in the Microsoft.Web provider, but there are no new errors and warnings introduced.

### CODE REVIEW

**Review by jianghaolu — CHANGES_REQUESTED** [review]

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json [hunk 1]
```diff
@@ -15,6 +15,236 @@
     "application/json"
   ],
   "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/hostingEnvironments/{name}/detectors": {
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: There are 4 new endpoints, 6 if you include the slot variations, and they all have examples. Let me know if you are not seeing one though.

**Comment by stevenernst131** [comment]
@jianghaolu I have added the richer descriptions and pushed the change.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: Sorry for the late response - You need to reference them through "x-ms-examples" on the operation. https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/x-ms-examples.md

[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: Hi @jianghaolu, thanks for the resource, I have added the references to the examples.

[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: Thanks @stevenernst131 I'll take a look at the examples now

**Review by jianghaolu — CHANGES_REQUESTED** [review]

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        "subscriptionId": "34adfa4f-cedf-4dc0-ba29-b6d1a69ab345",
+        "resourceGroupName": "Sample-WestUSResourceGroup",
+        "name": "SampleAppServiceEnvironment",
+        "detectorName": "runtimeavailability",
+        "api-version": "2016-03-01"
+    },
+    "responses": {
+        "200": {
+            "body": {
+                "value": {
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json]: Thanks for catching, I think this was a copy paste error. I have removed value in all 3 places

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListSiteDetectorResponses.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                        "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/Sample-WestUSResourceGroup/providers/Microsoft.Web/sites/SampleApp/detectors/runtimeavailability",
+                        "name": "runtimeavailability",
+                        "properties": {
+                            "metadata": {
+                                "description": "This detector analyzes the requests to your application.",
+                                "category": "Availability and Performance",
+                                "subCategory": "App Availability",
+                                "supportTopicId": ""
+                            },
+                            "dataset": [
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListSiteDetectorResponses.json]: dataset is an array, I have updated it in the definition.  It is an array in all examples.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        "subscriptionId": "34adfa4f-cedf-4dc0-ba29-b6d1a69ab345",
+        "resourceGroupName": "Sample-WestUSResourceGroup",
+        "siteName": "SampleApp",
+        "detectorName": "runtimeavailability",
+        "api-version": "2016-03-01"
+    },
+    "responses": {
+        "200": {
+            "body": {
+                "value": {
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json]: done

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListSiteDetectorResponsesSlot.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+                        "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/Sample-WestUSResourceGroup/providers/Microsoft.Web/sites/SampleApp/slots/staging/detectors/runtimeavailability",
+                        "name": "runtimeavailability",
+                        "properties": {
+                            "metadata": {
+                                "description": "This detector analyzes the requests to your application.",
+                                "category": "Availability and Performance",
+                                "subCategory": "App Availability",
+                                "supportTopicId": ""
+                            },
+                            "dataset": [
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListSiteDetectorResponsesSlot.json]: Changed to list.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponseSlot.json [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        "resourceGroupName": "Sample-WestUSResourceGroup",
+        "siteName": "SampleApp",
+        "slot": "staging",
+        "detectorName": "runtimeavailability",
+        "api-version": "2016-03-01"
+    },
+    "responses": {
+        "200": {
+            "body": {
+                "value": {
```
[the remark this answers is not in the document]
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponseSlot.json]: Done

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListHostingEnvironmentDetectorResponses.json [hunk 7]
```diff
@@ -8,22 +8,17 @@
...
-                                "category": "Availability and Performance",
-                                "subCategory": "App Availability",
-                                "supportTopicId": ""
-                            },
-                            "dataset": [
-                            ]
-                        }
-                    }
-                ]
+                "id": "/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/resourceGroups/Sample-WestUSResourceGroup/providers/Microsoft.Web/hostingEnvironments/SampleAppServiceEnvironment/detectors/runtimeavailability",
```
Comment (jianghaolu) [inline specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListHostingEnvironmentDetectorResponses.json]: the "value" here is actually correct since this is from a list operation.

**Review by jianghaolu — CHANGES_REQUESTED** [review]

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json]: This one, however, should have "value" removed.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        "siteName": "SampleApp",
+        "detectorName": "runtimeavailability",
+        "api-version": "2016-03-01"
+    },
+    "responses": {
+        "200": {
+            "body": {
+                "id": "/subscriptions/ef90e930-9d7f-4a60-8a99-748e0eea69de/resourceGroups/Build2015DemoRG/providers/Microsoft.Web/sites/BuggyBakery/detectors/runtimeavailability",
+                "name": "runtimeavailability",
+                "location": "West US",
```
Comment (jianghaolu) [inline specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json]: "location" is not allowed for a proxy resource.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json [hunk 9]
```diff
@@ -1681,9 +1681,13 @@
               "description": "metadata for the detector"
             },
             "dataset": {
-              "$ref": "#/definitions/DiagnosticData",
-              "description": "Data Set"
+              "type": "array",
+              "description": "Data Set",
+              "items": {
+                "$ref": "#/definitions/Solution"
```
Comment (jianghaolu) [inline specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: Why did you change the item type from `DiagnosticData` to `Solution`?

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListHostingEnvironmentDetectorResponses.json (hunk 7 above)
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_ListHostingEnvironmentDetectorResponses.json]: Removed it from the wrong file, added it back to this file and removed it from Diagnostics_GetHostingEnvironmentDetectorResponse.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json (hunk 8 above)
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetSiteDetectorResponse.json]: Removed

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json (hunk 9 above)
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/Diagnostics.json]: Copy paste error, corrected.

**Comment by stevenernst131** [comment]
Hi @jianghaolu, I believe I have addressed all your issues. Let me know if there is anything else from your side to be addressed.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                                        "2018-03-27T01:10:00Z",
+                                        "784642",
+                                        "0",
+                                        "0",
+                                        "0"
+                                    ]
+                                ]
+                            },
+                            "renderingProperties": {
+                                "renderingType": 2,
```
Comment (jianghaolu) [inline specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json]: This is defined as an enum but this is an integer. Same elsewhere.

**Comment by jianghaolu** [comment]
@stevenernst131 Sorry for the late response. One more issue and we should be good to go. (The new issue was in Dataset which was invalid earlier so the tool didn't report it earlier)

**Comment by jianghaolu** [comment]
Oh I just realized you changed all the error response types - let me sync with Samer to see what to do here.

File: specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json (hunk 10 above)
Comment (stevenernst131) [reply specification/web/resource-manager/Microsoft.Web/stable/2016-03-01/examples/Diagnostics_GetHostingEnvironmentDetectorResponse.json]: I have changed it to return a string, which is the actual way that we return enums.

**Comment by stevenernst131** [comment]
Hi @jianghaolu, I did not mean to change the Error Response in a way that caused a breaking change, that was not my intention. I have changed it back to how it was before, so let me know if that is correct.

**Comment by stevenernst131** [comment]
@jianghaolu Please let me know if there is anything else I need to do.

**Comment by jianghaolu** [comment]
@stevenernst131 Hi Steve! Sorry for the late response. Been crazy busy with other deliverables. 

Actually, do you think you can just remove those added "default" sections? Our SDKs default to throw `CloudException` when a default section is not provided. Changing them to `CommonDefinitions.json#/definitions/DefaultErrorResponse` is also a breaking change.

**Comment by stevenernst131** [comment]
@jianghaolu Thanks for the response, I have pushed a change to remove all the default sections, let me know if that works for you.

**Review by jianghaolu — APPROVED** [review]

**Comment by jianghaolu** [comment]
Thanks! Merging!
